makeSetupHook: support depsTargetTargetPropagated (fix using hooks when strictDeps)

Artturin opened the pull request:
Closes https://github.com/NixOS/nixpkgs/issues/49132

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
makeSetupHook: support depsTargetTargetPropagated

Artturin committed:
stdenv: allow propagating propagated dependencies separately from the
fixup phase

for makeSetupHook

SuperSandro2000 commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -65,6 +72,15 @@ makeSetupHook {
         ''
       );
 
+      basic-contains-gdk-pixbuf = let
+        tested = basic;
+      in testLib.runTest "basic-contains-gdk-pixbuf" (
+        testLib.skip stdenv.isDarwin ''
+          ${expectSomeLineContainingYInFileXToMentionZ "${tested}/bin/foo" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
+          ${expectSomeLineContainingYInFileXToMentionZ "${tested}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
```
```suggestion
      basic-contains-gdk-pixbuf = testLib.runTest "basic-contains-gdk-pixbuf" (
        testLib.skip stdenv.isDarwin ''
          ${expectSomeLineContainingYInFileXToMentionZ "${basic}/bin/foo" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
          ${expectSomeLineContainingYInFileXToMentionZ "${basic}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${lib.getLib librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"}
```

Artturin replied:
@jtojnar what do you think? since you added these tests

jtojnar commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -855,6 +855,39 @@ stripHash() {
 }
 
 
+propagatePropagatedDependencies() {
```
Maybe something like `{persist,record,output}PropagatedDependencies` so that we are not propagating what is already propagated?

Artturin replied:
changed to `recordPropagatedDependencies`

jtojnar commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -1200,35 +1233,7 @@ fixupPhase() {
 
 
     # Propagate dependencies & setup hook into the development output.
-    declare -ra flatVars=(
```
Rest of the `fixupPhase` function does not seem to use any of these variables (`substituteAll` only captures environment variables) so this looks good to me.

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
     makeWrapper
   ];
+
+  # depsTargetTargetPropagated will be buildInputs
```
Was this meant to be something like

```suggestion
  # depsTargetTargetPropagated will essentially be buildInputs when wrapGAppsHook is placed into nativeBuildInputs
```

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
   ];
+
+  # depsTargetTargetPropagated will be buildInputs
+  # the librsvg above should be removed but kept to not break anything that implicitly depended on its binaries
```
Maybe move the comment above then. I was quite confused until I noticed it.

Personally, I would just rip the band-aid off. I would not expect this to be very prevalent.

Artturin replied:
> Maybe move the comment above then. I was quite confused until I noticed it.

this comment and the librsvg above it are closer now after rearranging the list

> Personally, I would just rip the band-aid off. I would not expect this to be very prevalent.

Should be a separate PR

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -27,16 +28,22 @@ makeSetupHook {
+    # SVG icons. Most icon themes are SVG-based and so are some
+    # graphics in GTK (e.g. cross for closing window in window title bar)
+    # so it is pretty much required for applications using GTK.
+    librsvg
```
This should probably go to a separate commit from “makeSetupHook: support depsTargetTargetPropagated”

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -559,10 +559,22 @@ rec {
-  makeSetupHook = { name ? "hook", deps ? [], substitutions ? {}, meta ? {}, passthru ? {} }: script:
+  makeSetupHook =
+    { name ? "hook"
+    , substitutions ? { }
+    , meta ? { }
+    , passthru ? { }
+    , deps ? [ ]
```
Why change the ordering? It does not appear to be alphabetical nor does it follow the [common order for `mkDerivation`](https://github.com/jtojnar/nixpkgs-hammering/blob/1b038ef38fececb39b65a4cdfa7273ed9d9359b4/explanations/attribute-ordering.md).

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -559,10 +559,22 @@ rec {
+    , propagatedBuildInputs ? [ ]
+    # these will be buildInputs
+    , depsTargetTargetPropagated ? [ ]
+    }: script:
```
Improve legibility:

```suggestion
    }:
    script:
```

jtojnar commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -559,10 +559,22 @@ rec {
       (substitutions // {
         inherit meta;
+        inherit depsTargetTargetPropagated;
+        propagatedBuildInputs = assert (lib.assertMsg (lib.isList deps) "deps argument to makeSetupHook must be a list. content: ${toString deps}"); deps ++ propagatedBuildInputs;
```
I would go for a soft deprecation (using a `trace`) first so that external consumers can adapt.

And while at it we might as well deprecate `deps` in favour of `propagatedBuildInputs`.

Artturin replied:
> I would go for a soft deprecation (using a trace) first so that external consumers can adapt.

done

Artturin replied:
and deps deprecated

jtojnar reviewed (commented):
Nice work.

Artturin committed:
wrapGAppsHook: make it work with strictDeps

remove unused input

add test for GDK_PIXBUF_MODULE_FILE

run tests with strictDeps

if the tests work with strictDeps then they work with non strictDeps

Artturin committed:
treewide: makeSetupHook deps -> propagatedBuildInputs

Artturin committed:
makeSetupHook: deprecate deps argument

Artturin committed:
add docs for makeSetupHook

winterqt commented:
Should every instance of e.g. Cargo and pip be from buildPackages, as makeWrapper is?

Artturin commented:
> Should every instance of e.g. Cargo and pip be from buildPackages, as makeWrapper is?

No, thats just a workaround for hooks in buildInputs / hook in nativeBuildInputs but splicing not working
